### PR TITLE
Statistical convection model solver (solve_model)

### PR DESCRIPTION
**Pull request by egthomas**
This pull request adds a new `solve_model` binary which, given solar wind / geomagnetic conditions as input via the command line, solves for the model potential and velocity values over either a uniform or equal-area grid and writes to stdout in a fixed-column ASCII-style output.  This supports all available SuperDARN convection models in the RST (RG96, PSR10, CS10, TS18, TS18-Kp).  Below is sample output:

```
Date:  2018-05-21 00:00
Model: TS18-Kp
Bin:   6<Kp<8, Bz-
Grid:  Equal-area (lat_step: 1.00 [deg])

MLAT [deg]   MLT [hr]   Pot [kV] Vazm [deg] Vmag [m/s]
---------- ---------- ---------- ---------- ----------
   46.5000  0.0483871  -0.138252   -90.2309    63.7825
   46.5000   0.145161  -0.136914   -90.2490    63.0686
   46.5000   0.241935  -0.135492   -90.2674    62.3145
   46.5000   0.338710  -0.133989   -90.2861    61.5213
   46.5000   0.435484  -0.132407   -90.3051    60.6902
   46.5000   0.532258  -0.130747   -90.3244    59.8223
...
```
And using a simple IDL routine to read and plot the model velocity vectors and potential contours:

image

Because this program relies on TS18 and TS18-Kp model functionality I am requesting to merge it into the `feature/ts18_model` branch first rather than directly to `develop`.

### CODE REVIEW

**Comment by ksterne** [comment]
hey @egthomas, since #138 is dependent on this one, trying to take a quick look at it.  Probably don't have time to write the simple read/plot code so just going to see if things make some kind of sense here.  One thing in using it, it wasn't apparent in the `--help` output that setting the date and time was possible since it's not in the first couple of lines that list some of the options.  Maybe this is me being dumb, but that might be because those parameters aren't required?  But if that's the case, I left off `-by` and it still ran without any kind of warning message.  It doesn't note what the default By value might be.

In fact, this runs without giving any kind of solar wind or geomagnetic conditions...so I think your first sentence should read "given nothing on command line, solves.."  Is this the expected behavior you were thinking for this or should more errors/warnings about default values be added?

**Comment by egthomas** [comment]
It works the same way as `map_addmodel` - if you don't provide any solar wind values as input then it will set everything to zero by default.

**Comment by ksterne** [comment]
That sounds good, think it's worth noting the defaults in the help message?

**Comment by egthomas** [comment]
Below is a simple IDL program which will read the output ASCII file and plot the model potential contours (for some reason github wouldn't let me upload something with a .pro extension):

```
; This program reads the ASCII output from the solve_model
; binary in the RST and plots the model potential contours 
; on an MLAT/MLT grid - EGT 20180527

pro solve_model_rst, fname

; Check to make sure one filename was provided
if n_params() ne 1 then begin
    print, 'Must provide filename, returning'
    return
endif

if ~file_test(fname) then begin
    print, 'File '+fname+' does not exist, returning'
    return
endif

; Get number of lines in file
nrecs = file_lines(fname)

; Set up arrays to read header and data
hdr = strarr(7)
cnt = 0L
mlat = fltarr(nrecs)
mlt = fltarr(nrecs)
pot = fltarr(nrecs)
azm = fltarr(nrecs)
mag = fltarr(nrecs)

; Open model text file
openr, ilun, fname, /get_lun

; Read header lines
readf, ilun, hdr

; Read all records in file
while ~eof(ilun) do begin
    readf, ilun, amlat, amlt, apot, aazm, amag
    mlat[cnt] = abs(amlat)
    mlt[cnt] = amlt
    pot[cnt] = apot
    azm[cnt] = aazm
    mag[cnt] = amag
    cnt += 1L
endwhile

; Close model text file
free_lun, ilun

; Trim arrays to number of records
mlat = mlat[0:cnt-1L]
mlt = mlt[0:cnt-1L]
pot = pot[0:cnt-1L]
azm = azm[0:cnt-1L]
mag = mag[0:cnt-1L]

; Set up plotting area
plot, [0,0], /nodata, xrange=[-50,50], yrange=[-50,50], /xstyle, /ystyle, /iso

; Convert MLAT/MLT to polar coords for contouring
polarx = fltarr(cnt)
polary = fltarr(cnt)

for ii=0L,cnt-1L do begin
    polarx[ii] = (90.-abs(mlat[ii]))*sin(mlt[ii]*15.0*!pi/180.)
    polary[ii] = (90.-abs(mlat[ii]))*cos(mlt[ii]*15.0*!pi/180.)*(-1.)
endfor

; Set some options for drawing potential contours
n_levels = 10
diffc = 6
symsize = 2.

; Overlay negative contours
contour, pot, polarx, polary, /overplot, xstyle=4, ystyle=4, $
    levels=-57+findgen(n_levels)*diffc, max_value=100, $
    /follow, noclip=0, /irregular

; Overlay positive contours
contour, pot, polarx, polary, /overplot, xstyle=4, ystyle=4, $
    levels=3+findgen(n_levels)*diffc, max_value=100, $
    /follow, noclip=0, /irregular

; Overlay symbols at potential minima/maxima
pot_max = max(pot, maxind, min=pot_min, subscript_min=minind)
usersym, [-1., 1.0], [0., 0.], thick=2.
plots, polarx[maxind], polary[maxind], psym=1, $
    symsize=0.75*symsize, noclip=0
plots, polarx[minind], polary[minind], psym=8, $
    symsize=0.75*symsize, noclip=0

end
```

This example was tested with IDL 8.5+ but it should be backwards compatible with earlier versions.  My code to draw the model velocity vectors was too dependent on custom/local functions to include in this example.

**Comment by ksterne** [comment]
Looks like the plotting code you put here doesn't plot the velocity vectors, right?  Just wanted to make sure.

**Comment by ksterne** [comment]
So, I think there's something off here on maybe one very small use case.  Whenever I use: `solve_model -equal -psr10 -tilt -10 -bz 1 -by 0` I get an output that follows the negative tilt.  However, if I look at the southern hemisphere with: `solve_model -sh -equal -psr10 -tilt -10 -bz 1 -by 0` I get an output that says neutral tilt and matches the neutral tilt on: http://superdarn.thayer.dartmouth.edu/models/PSR10.html.  Is the `tilt` option being overwritten or not read with the `-sh` option?  Or is this supposed to happen (not being familiar with models)?

**Comment by egthomas** [comment]
The sign of dipole tilt is reversed for the Southern Hemisphere.  The boundary between neutral/negative tilt is -10 degrees, so I can't remember which option it selects if you input a dipole tilt angle exactly on that boundary between model bins.

Here is an excerpt from [Pettigrew et al. [2010]](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2009JA014956) explaining the dipole tilt angle convention (which has been maintained for the CS10 and TS18 models):

> Furthermore, in this paper the sign of the dipole tilt angle is switched in the Southern Hemisphere to allow direct comparison with the Northern Hemisphere. We use this sign‐adjusted dipole tilt angle, hereafter referred to simply as tilt, to sort our data since it is a more accurate measure of the Earth's magnetic field geometry and of the amount of solar‐produced ionospheric conductivity (possible sources of the observed dependence). The tilt is calculated using the International Geomagnetic Reference Field (IGRF) model [Mandea and Macmillan, 2000] and the data are classified as: negative tilt (tilt < −10°), neutral tilt (−10° < tilt <10°), or positive tilt (tilt >10°).

**Comment by ksterne** [comment]
Ha, that's funny that I picked exactly the boundary value without even knowing it.  That makes sense too that things are flipped in the southern hemisphere.  Thanks for the explanation!  

And whoops, sorry I glossed over the last sentence you had on plotting the velocity vectors.  Missed that at the end of the long block of code.

OK, in more testing (and probably more that I don't know about this stuff), I tried pushing the tilt just over the edge with `solve_model -sh -equal -psr10 -tilt 10.1 -bz 1 -by 0` and got a seg fault.  Am I giving it bad parameters?  It seems as though anything less than -10 for tilt causes a seg fault here as I've tried -15 and -20, so it's not just an integer/float problem.

**Comment by egthomas** [comment]
No, that shouldn't be happening.  I can replicate the seg fault with the command `solve_model -sh -psr10 -tilt -15`, and think I've traced it to the fact that certain positive dipole patterns don't exist for the PSR10 model (eg, 0-3 nT and positive dipole tilt which is undefined in the SH for all clock angle orientations).

The solution will be to add better error checking for these PSR10 patterns in both the `solve_model` and `map_addmodel` binaries.

File: codebase/superdarn/src.bin/tk/tool/solve_model.1.0/doc/solve_model.doc.xml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+</option>
+<option><on>-bz <ar>bz</ar></on><od>set the Z component of the magnetic field to <ar>bz</ar> [nT]. The default value is 0 nT.</od>
+</option>
+<option><on>-vx <ar>vx</ar></on><od>set the solar wind velocity to <ar>vx</ar> [km/s]. The default value is 450 km/s.</od>
+</option>
+<option><on>-tilt <ar>tilt</ar></on><od>set the dipole tilt angle to <ar>tilt</ar> [degrees]. The default value is determined by date/time.</od>
+</option>
+<option><on>-kp <ar>kp</ar></on><od>set the Kp magnetic index to <ar>kp</ar>. The default value is 0.</od>
+</option>
+<option><on>-d <ar>yyymmdd</ar></on><od>calculate for the date <ar>yyyymmdd</ar>. The default value is the current date.</od>
```
Comment (asreimer) [inline codebase/superdarn/src.bin/tk/tool/solve_model.1.0/doc/solve_model.doc.xml]: yyymmdd should be yyyymmdd

**Comment by asreimer** [comment]
I tried testing the code and the following produces a seg fault:

    solve_model -bz 10 -d 20181112 -t 05:20 -ts18

Am I doing something stupid? It works if I don't specify the date:

    solve_model -bz 10 -t 05:20 -ts18

**Comment by asreimer** [comment]
I've had a look at the code and then only thing I can see that might be causing the segfault is something to do with the strdate function or the lines near it.

https://github.com/SuperDARN/rst/blob/c3b4bc685b1f11b5d1252c2ba9322ec3877444ca/codebase/superdarn/src.bin/tk/tool/solve_model.1.0/solve_model.c#L311-L323

**Comment by egthomas** [comment]
@asreimer - thanks for catching those issues.  I think the seg fault when using the date option should now be fixed.

**Comment by asreimer** [comment]
@egthomas oh that's a nasty little typo.

Thanks for fixing this so promptly! I am confirming that the segfault is no longer an issue with regards to the date option. I'm working on a shell script that can throughly test this code. When it's done, I'll post it. The intent is just to cycle through various combinations of inputs and see if anything segfaults.

**Comment by asreimer** [comment]
If you run the following code:

    solve_model -bz -5 -by 5 -vx 600 -d 20181112 -t 0520 -ts18 > output.txt

And then use [this python plotting code that I derived](https://github.com/SuperDARN/rst/files/2107846/plot_potential.txt) from @egthomas IDL code (rename the file to .py), you should get this figure:


figure_1-6

Run the python code with:

    python plot_potential.py output.txt

**Comment by egthomas** [comment]
@asreimer that figure looks reasonable.  Just keep in mind the PSR10 issue pointed out by @ksterne when cycling through available patterns.

**Comment by asreimer** [comment]
@egthomas @ksterne I don't get a segfault for `solve_model -sh -equal -psr10 -tilt 10.1 -bz 1 -by 0`

I'm supposed to?

**Comment by asreimer** [comment]
@egthomas @ksterne I don't get a segfault for `solve_model -sh -equal -psr10 -tilt 10.1 -bz 1 -by 0`

I'm supposed to?

**Comment by asreimer** [comment]
Ah nevermind, it's `solve_model -sh -psr10 -tilt -15` that causes the segfault. Ok.

**Comment by asreimer** [comment]
Alright, here's my testing scripts:
[solve_model_tests_multiproc.txt](https://github.com/SuperDARN/rst/files/2112670/solve_model_tests_multiproc.txt)
[plot_potential.txt](https://github.com/SuperDARN/rst/files/2112669/plot_potential.txt)

You need to rename both of these files to have a .py extension (github only let's you attach .txt files). Put them in the same directory and then edit the `solve_model_tests_multiproc.py` file. I recommend turning plots off (should be off by default). If you don't, it could take about 50 days to run all of these tests (single threaded) and consume ~850 GB of space. This can be reduced if you tweak the number of bx, by, vx, etc that are tested. You can also using multiprocessing (enabled with 30 threads by default) and then change the number of threads to use.

I'm currently running this on 30 threads. It's probably overkill, but I'm just interested to see if anything happens. It should be done running tomorrow.

**Comment by asreimer** [comment]
In summary, here's all of the options that I'm running permutations of:

    bys = ['', '-by -12 ', '-by -8 ', '-by -4 ', '-by 0 ', '-by 4 ', '-by 8 ', '-by 12 ']
    bzs = ['', '-bz -13 ', '-bz -9 ', '-bz -5 ', '-bz -1 ', '-bz 3 ', '-bz 7 ', '-bz 11 ']
    vxs = ['', '-vx 496 ', '-vx 222 ', '-vx 821 ', '-vx 1399 ']
    tilts = ['', '-tilt 1 ', '-tilt 3 ', '-tilt -2 ']
    kps = ['', '-kp 1 ', '-kp 2 ', '-kp 3 ', '-kp 4 ', '-kp 5 ', '-kp 6 ', '-kp 7 ', '-kp 8 ']
    dates = ['', '-d 20180110 ', '-d 20010203 ', '-d 19990417 ', '-d 20200822 ']
    times = ['', '-t 0000 ', '-t 0101 ', '-t 2028 ']
    models = ['-rg96 ', '-psr10 ', '-cs10 ', '-ts18 ', '-ts18_kp ']
    hemis = ['', '-sh ']

That's 2,304,000 permutations.

**Comment by egthomas** [comment]
@asreimer if you wanted you could cut down on the number of permutations by only applying some of the options to certain models.  For example, `-kp` input is only used with the TS18-Kp model, `-vx` is only used with the CS10 and TS18 models, etc.

**Comment by asreimer** [comment]
@egthomas It's true, but that would have taken longer for me to program in to the tests, so I didn't bother.

Results are in: All permutations worked, except for 2070 of them where psr10 was called with the default tilt angle causing it to segfault.

I'd say this branch is easily ready to be merged. Do I have permission?

**Comment by asreimer** [comment]
Maybe another bug here. This was found by our summer student.

The date that the model is run for or the date written to the output are wrong.

## reproduce the bug
Run:

    solve_model -bz -5 -by 5 -vx 600 -d 20161012 -t 0520 -ts18 > 20161012.txt

and then you should see:

    Date:  2016-11-02 16:00
    Model: TS18
    Bin:   Esw    4.2 mV/m, Bang   135 deg., tilt  -5.5 deg.
    Grid:  Uniform (lat_step: 1.00, lon_step: 2.00 [deg])
    
    MLAT [deg]   MLT [hr]   Pot [kV] Vazm [deg] Vmag [m/s]
    ---------- ---------- ---------- ---------- ----------
       53.5000    0.00000  -0.133294   -91.3427    57.5358

Notice the input data and time were `2016-10-12 05:20`, yet the date written to the file was `2016-11-02 16:00`

@egthomas Any ideas? I'll try to sort this out, but I don't have a lot of time this week.

**Comment by asreimer** [comment]
I think the bug is related to the `-t` option. Specifically, if you run the command:

    solve_model -bz -5 -by 5 -vx 600 -d 20161012 -ts18 > 20161012.txt

## AHA

Turns out if you don't include a __:__ in the time then the code is happy to run but gives you the wrong answer.

So I recommend some input checking if possible here. The documentation correctly says that you need to do `-t hh:mm` but if you don't, nothing gets mad at you.

**Comment by asreimer** [comment]
@egthomas That looks great to me. What's the procedure for merging? Should we get @ksterne or someone else to have another look?

I honestly think this is good to go now. It was good to go before I made the user input error even.

P.S. I do have edit permissions now, so I can merge this.

**Comment by egthomas** [comment]
I still need to add some kind of error checking for the missing PSR10/CS10 patterns, but we're in the process of moving this week so my time is very limited.  Of course this pull request is aimed at the `feature/ts18_model` branch so I could still make the fix there before merging into `develop`.

**Comment by asreimer** [comment]
Focus on moving. This isn't time critical since it's working and we can just run things from this branch in the mean time. I think it's worth waiting for the error checking for missing PSR10/CS10 patterns since that's more related to the solve_model code than the ts18 model.

**Comment by egthomas** [comment]
I've added a check for missing/undefined PSR10 patterns to the `solve_model` code so this branch should now be ready to merge.

**Comment by ksterne** [comment]
Yep, thanks @egthomas for adding the check for places where PSR10 isn't defined and good error message.  Should make this a little bit more user friendly for people that aren't as familiar or ones that do understand it but make a typo.
